lib/gigo/active_record/base.rb

```diff
@@ -1,107 +1,77 @@
-              def #{attr}
-                begin
-                  GIGO.load(super)
-                rescue NoMethodError, ActiveModel::MissingAttributeError
-                  nil
-                end
+        end
+        attrs.each do |attr|
+          mod.module_eval <<-CODE, __FILE__, __LINE__
+            def #{attr}
+              begin
+                GIGO.load(super)
+              rescue NoMethodError, ActiveModel::MissingAttributeError
+                nil
               end
-            CODE
-          end
+            end
+          CODE
         end
+      end
 
-        def gigo_columns(*excepts)
-          cols = columns.select{ |c| c.type == :string || c.type == :text }.map{ |c| c.name.to_s } - excepts.map(&:to_s)
-          gigo_column(*cols)
-        end
+      def gigo_columns(*excepts)
+        cols = columns.select{ |c| c.type == :string || c.type == :text }.map{ |c| c.name.to_s } - excepts.map(&:to_s)
+        gigo_column(*cols)
+      end
 
-        def after_initialize_gigoize_attributes
-          after_initialize :gigoize_attributes
-          define_method :gigoize_attributes do
-            self.class.const_get(:GIGOColumns).instance_methods.each { |name| self.send(name) }
-          end
-          private :gigoize_attributes
+      def after_initialize_gigoize_attributes
+        after_initialize :gigoize_attributes
+        define_method :gigoize_attributes do
+          self.class.const_get(:GIGOColumns).instance_methods.each { |name| self.send(name) }
         end
-
+        private :gigoize_attributes
       end
 
-      module ThreeOhOnly
+      def gigo_coder_for(klass)
+        GigoCoder.new(klass)
+      end
 
-        include Shared
+      class GigoCoder
+        attr_reader :klass
 
-        def gigo_serialized_attribute(*attrs)
-          # ActiveRecord 3.0 has no proper hooks. So we fix all YAML object laoding for this class.
-          include InstanceMethods
+        def initialize(klass)
+          @klass = klass
         end
 
-        module InstanceMethods
-
-          private
-
-          def object_from_yaml(string)
-            return string unless string.is_a?(String) && string =~ /^---/
-            begin
-              existing_encoding = Encoding.default_internal
-              Encoding.default_internal = GIGO.encoding
-              s = GIGO.load(string)
-              YAML::load(s) rescue s
-            ensure
-              Encoding.default_internal = existing_encoding
-            end
+        def load(yaml)
+          return klass.new if yaml.nil?
+          existing_encoding = Encoding.default_internal
+          Encoding.default_internal = GIGO.encoding
+          value = YAML.load(GIGO.load(yaml))
+          unless value.is_a?(klass)
+            raise SerializationTypeMismatch, "Attribute was supposed to be a #{klass.to_s}, but was a #{hash.class}."
           end
-
+          value
+        ensure
+          Encoding.default_internal = existing_encoding
```

Comment:
Great use of ensure here! But do you need a check here because of the 1st line guard clause? IE, this ensure would run but the local variable `existing_encoding` would be nil or worse undefined. So maybe use an instance variable or something here? Like...

```ruby
      class GigoCoder
        attr_reader :klass

        def initialize(klass)
          @klass = klass
          @default_internal = Encoding.default_internal
        end

        def load(yaml)
          # ...
        ensure
          Encoding.default_internal = @default_internal
        end
      end
```